lib/analytics.js

```diff
@@ -678,22 +694,39 @@ Analytics.prototype._invoke = function(method, facade) {
 
   var failedInitializations = self.failedInitializations || [];
   each(function(integration, name) {
-    if (!facade.enabled(name)) return;
+    var facadeCopy = extend(true, {}, facade);
+
+    if (!facadeCopy.enabled(name)) return;
     // Check if an integration failed to initialize.
     // If so, do not process the message as the integration is in an unstable state.
     if (failedInitializations.indexOf(name) >= 0) {
       self.log(
-        'Skipping invokation of .%s method of %s integration. Integation failed to initialize properly.',
+        'Skipping invocation of .%s method of %s integration. Integration failed to initialize properly.',
         method,
         name
       );
     } else {
       try {
-        metrics.increment('analytics_js.integration.invoke', {
-          method: method,
-          integration_name: integration.name
+        // Apply any integration middlewares that exist, then invoke the integration with the result.
+        self._integrationMiddlewares.applyMiddlewares(facadeCopy.obj, function(
+          payload
+        ) {
+          // A nullified payload should not be sent to an integration.
+          if (payload === null) {
+            self.log(
+              '' + 'Payload to integration "%s" was null and dropped.',
+              name
+            );
+            return;
+          }
+          facadeCopy.obj = payload;
```

Comment:
can we create a new facade instance, instead of mutating the existing one?

Reply:
We're already using a copy - should this be fine? (Not sure if a copy of a copy is necessary here)

Reply:
Also not sure if generating a new facade will or will not create further mutations that might not be expected.

Reply:
Ideally I would like to treat these as if they were immutable.

Currently the copy is before any middlewares are called. I think we should make a new copy after  we get the result of the middlewares.

Unless this degrades the performance significantly.

Reply:
Sure, I think that's a good enough argument then.